Claim data migration

carolynvs opened the pull request:
# What does this change
Migrates claim data to the new layout (splitting results and outputs from the claim).

Example Output

```console
$ porter list
!!! Migrating claims data to match the CNAB Claim spec https://cdn.cnab.io/schema/cnab-claim-1.0.0-DRAFT+b5ed2f3/claim.schema.json. This is a one-way migration !!!
 - Migrating claim asciiart to the new claim layout...
 - Migrating claim az-iot to the new claim layout...
 - Migrating claim azure-cli to the new claim layout...
 - Migrating claim cowsay to the new claim layout...
 - Migrating claim example-exec-outputs to the new claim layout...
 - Migrating claim gke-example to the new claim layout...
 - Migrating claim my-unique-demo to the new claim layout...
 - Migrating claim mycowsay to the new claim layout...
 - Migrating claim porter-wordpress-mysql to the new claim layout...
 - Migrating claim spring-music-bundle to the new claim layout...
 - Migrating claim spring-music to the new claim layout...
 - Migrating claim whalegap to the new claim layout...
!!! Migration complete !!!
NAME                     CREATED      MODIFIED     LAST ACTION   LAST STATUS
whalegap                 2020-06-12   2020-06-12   install       succeeded
porter-wordpress-mysql   2019-11-11   2019-11-11   install       succeeded
example-exec-outputs     2019-11-08   2019-11-08   install       succeeded
spring-music-bundle      2019-10-03   2019-10-03   install       failed
mycowsay                 2019-10-01   2019-10-01   install       succeeded
cowsay                   2019-10-01   2019-10-01   say           succeeded
gke-example              2019-09-27   2019-09-27   install       succeeded
spring-music             2019-09-26   2019-09-26   install       succeeded
my-unique-demo           2019-09-26   2019-09-26   install       failed
az-iot                   2019-09-11   2019-09-11   install       succeeded
asciiart                 2019-06-24   2019-06-24   install       succeeded
azure-cli                2019-06-07   2019-06-07   install       succeeded

$ porter list
NAME                     CREATED      MODIFIED     LAST ACTION   LAST STATUS
whalegap                 2020-06-12   2020-06-12   install       succeeded
porter-wordpress-mysql   2019-11-11   2019-11-11   install       succeeded
example-exec-outputs     2019-11-08   2019-11-08   install       succeeded
spring-music-bundle      2019-10-03   2019-10-03   install       failed
mycowsay                 2019-10-01   2019-10-01   install       succeeded
cowsay                   2019-10-01   2019-10-01   say           succeeded
gke-example              2019-09-27   2019-09-27   install       succeeded
spring-music             2019-09-26   2019-09-26   install       succeeded
my-unique-demo           2019-09-26   2019-09-26   install       failed
az-iot                   2019-09-11   2019-09-11   install       succeeded
asciiart                 2019-06-24   2019-06-24   install       succeeded
azure-cli                2019-06-07   2019-06-07   install       succeeded
```
# What issue does it fix
Closes #937 

# Notes for the reviewer
This only works for the local filesystem. The azure plugin needs to be updated to support the new crud.Store interface before it will work for claims stored in Azure.

🚨 Relies on #1065, which must be merged first

TODO:
- [x] schema.json has the schema for credentials in it too, should we keep that and populate it as well? right now the file looks like this 

```json
{
    "claims": "cnab-claim-1.0.0-DRAFT+b5ed2f3",
    "credentials": ""
}
```

# Checklist
- [x] Unit Tests
- [ ] Documentation ❓ Not sure it's needed
- [x] Schema (porter.yaml) - Not impacted

vdice commented:
> The azure plugin needs to be updated to support the new crud.Store interface before it will work for claims stored in Azure.

Should an issue in https://github.com/deislabs/porter-azure-plugins/issues be created?

For this PR, pending merge of #1065 , is https://github.com/deislabs/porter/pull/1090/commits/03e1272fea9286000e788cd7afae4ba3a07b2675 the main commit to look at?

vdice commented on pkg/claims/migrateClaimsWrapper.go:
```diff
@@ -35,33 +54,184 @@ func (w *migrateClaimsWrapper) Store() crud.Store {
+}
+
+func (w *migrateClaimsWrapper) MigrateAll() error {
+	fmt.Fprint(w.Err, "!!! Migrating claims data to match the CNAB Claim spec https://cdn.cnab.io/schema/cnab-claim-1.0.0-DRAFT+b5ed2f3/claim.schema.json. This is a one-way migration !!!\n")
```
Should we Fprintf using `claim.CNABSpecVersion` instead of the hard-coded version here?

carolynvs replied:
This migration is specific to getting them from the before times to cnab-claim-1.0.0-DRAFT+d7ffba8. When we bump claims down the road, I think we may have a different migration. I am not sure we want to automatically tie this together just yet since it's not a generic migration.

vdice commented on pkg/claims/migrateClaimsWrapper.go:
```diff
@@ -35,33 +54,184 @@ func (w *migrateClaimsWrapper) Store() crud.Store {
-	// If we can't migrate the data at any point, print an error message and
-	// return the original claim so that the entire operation isn't halted for
-	// remaining claims
+	fmt.Fprintf(w.Err, "!!! Migration complete !!!\n")
```
Thoughts on only printing the migration details (here and elsewhere) if `porter-debug` is true?

carolynvs replied:
We used to print on  --debug but this is a critical migration that could go wrong and it's important for people to see it when it runs (it only runs once). If something goes wrong, you can't rerun with --debug on, you've missed it at that point.

carolynvs commented:
Let me push up my latest work but it's still a WIP (both for this and the azure plugin). I've added a schema.json file to the root of the home directory that says what schema version the layout follows, and am using it to tell if we should migrate. It works for a filesystem plugin. 

It isn't working for my implementation of the azure plugin because I changed how we query for blobs (by tags) and the migration relies on querying instead of listing. So I need to think a bit more...

carolynvs commented:
@vdice This is ready to review when you have time.
